File: machine/ts/machine/memory/lua_heap_usage.ts
import { RAM_SIZE } from './map';

type LuaHeapUsageHooks = {
	context: unknown;
	getBaseRamUsedBytes(context: unknown): number;
	collectTrackedHeapBytes(context: unknown): number;
};

const MIN_COLLECTION_BYTES = 1024 * 1024;
const DEFAULT_LUA_HEAP_USAGE_CONTEXT = 0;

let trackedLuaHeapBytes = 0;
let nextCollectionBytes = MIN_COLLECTION_BYTES;
let pendingLuaHeapCollection = false;
let luaHeapUsageHooks: LuaHeapUsageHooks = {
	context: DEFAULT_LUA_HEAP_USAGE_CONTEXT,
	getBaseRamUsedBytes: zeroBaseRamUsedBytes,
	collectTrackedHeapBytes: zeroTrackedHeapBytes,
};

function zeroBaseRamUsedBytes(_context: unknown): number {
	return 0;
}

function zeroTrackedHeapBytes(_context: unknown): number {
	return trackedLuaHeapBytes;
}

export function configureLuaHeapUsage<TContext>(
	context: TContext,
	getBaseRamUsedBytes: (context: TContext) => number,
	collectTrackedHeapBytes: (context: TContext) => number,
): void {
	luaHeapUsageHooks = {
		context,
		getBaseRamUsedBytes: getBaseRamUsedBytes as (context: unknown) => number,
		collectTrackedHeapBytes: collectTrackedHeapBytes as (context: unknown) => number,
	};
}

export function resetLuaHeapUsageHooks(): void {
	luaHeapUsageHooks = {
		context: DEFAULT_LUA_HEAP_USAGE_CONTEXT,
		getBaseRamUsedBytes: zeroBaseRamUsedBytes,
		collectTrackedHeapBytes: zeroTrackedHeapBytes,
	};
}

export function resetTrackedLuaHeapBytes(): void {
	trackedLuaHeapBytes = 0;
	nextCollectionBytes = MIN_COLLECTION_BYTES;
	pendingLuaHeapCollection = false;
}

export function addTrackedLuaHeapBytes(delta: number): void {
	trackedLuaHeapBytes += delta;
	if (trackedLuaHeapBytes < 0) {
		throw new Error('[LuaHeapUsage] Tracked heap bytes underflow.');
	}
	if (delta > 0) {
		requestLuaHeapCollectionIfNeeded();
	}
}

export function getTrackedLuaHeapBytes(): number {
	return trackedLuaHeapBytes;
}

export function enforceLuaHeapBudget(): void {
	const totalRamUsedBytes = luaHeapUsageHooks.getBaseRamUsedBytes(luaHeapUsageHooks.context) + trackedLuaHeapBytes;
	if (!pendingLuaHeapCollection && totalRamUsedBytes < RAM_SIZE) {
		return;
	}
	collectTrackedLuaHeapBytes();
}

export function collectTrackedLuaHeapBytes(): void {
	pendingLuaHeapCollection = false;
	trackedLuaHeapBytes = luaHeapUsageHooks.collectTrackedHeapBytes(luaHeapUsageHooks.context);
	nextCollectionBytes = Math.max(MIN_COLLECTION_BYTES, trackedLuaHeapBytes * 2);
	const totalRamUsedBytes = luaHeapUsageHooks.getBaseRamUsedBytes(luaHeapUsageHooks.context) + trackedLuaHeapBytes;
	if (totalRamUsedBytes >= RAM_SIZE) {
		throw new Error(`[LuaHeap] Out of heap memory (${totalRamUsedBytes} >= ${RAM_SIZE}).`);
	}
}

function requestLuaHeapCollectionIfNeeded(): void {
	const totalRamUsedBytes = luaHeapUsageHooks.getBaseRamUsedBytes(luaHeapUsageHooks.context) + trackedLuaHeapBytes;
	if (trackedLuaHeapBytes <= nextCollectionBytes && totalRamUsedBytes < RAM_SIZE) {
		return;
	}
	pendingLuaHeapCollection = true;
}

File: machine/cpp/machine/memory/lua_heap_usage.cpp
#include "machine/memory/lua_heap_usage.h"
#include "machine/memory/map.h"
#include <algorithm>
#include <cstdint>
#include <stdexcept>
#include <string>

namespace bmsx {
namespace {

constexpr size_t MIN_COLLECTION_BYTES = 1024 * 1024;

size_t g_trackedLuaHeapBytes = 0;
size_t g_nextCollectionBytes = MIN_COLLECTION_BYTES;
bool g_pendingLuaHeapCollection = false;
uint8_t g_defaultLuaHeapUsageContext = 0;

size_t zeroBaseRamUsedBytes([[maybe_unused]] void* context) {
	return 0;
}

size_t zeroTrackedHeapBytes([[maybe_unused]] void* context) {
	return g_trackedLuaHeapBytes;
}

LuaHeapUsageHooks g_luaHeapUsageHooks = {
	&g_defaultLuaHeapUsageContext,
	&zeroBaseRamUsedBytes,
	&zeroTrackedHeapBytes,
};

void collectTrackedLuaHeapBytesNow() {
	g_pendingLuaHeapCollection = false;
	g_trackedLuaHeapBytes = g_luaHeapUsageHooks.collectTrackedHeapBytes(g_luaHeapUsageHooks.context);
	g_nextCollectionBytes = std::max(MIN_COLLECTION_BYTES, g_trackedLuaHeapBytes * 2);
	const size_t totalRamUsedBytes = g_luaHeapUsageHooks.getBaseRamUsedBytes(g_luaHeapUsageHooks.context) + g_trackedLuaHeapBytes;
	if (totalRamUsedBytes >= RAM_SIZE) {
		throw std::runtime_error("[LuaHeap] Out of heap memory (" + std::to_string(totalRamUsedBytes) + " >= " + std::to_string(RAM_SIZE) + ").");
	}
}

void requestLuaHeapCollectionIfNeeded() {
	const size_t totalRamUsedBytes = g_luaHeapUsageHooks.getBaseRamUsedBytes(g_luaHeapUsageHooks.context) + g_trackedLuaHeapBytes;
	if (g_trackedLuaHeapBytes <= g_nextCollectionBytes && totalRamUsedBytes < RAM_SIZE) {
		return;
	}
	g_pendingLuaHeapCollection = true;
}

}

void configureLuaHeapUsage(void* context, LuaHeapUsageReader getBaseRamUsedBytes, LuaHeapUsageReader collectTrackedHeapBytes) {
	g_luaHeapUsageHooks = {
		context,
		getBaseRamUsedBytes,
		collectTrackedHeapBytes,
	};
}

void resetLuaHeapUsageHooks() {
	g_luaHeapUsageHooks = {
		&g_defaultLuaHeapUsageContext,
		&zeroBaseRamUsedBytes,
		&zeroTrackedHeapBytes,
	};
}

void resetTrackedLuaHeapBytes() {
	g_trackedLuaHeapBytes = 0;
	g_nextCollectionBytes = MIN_COLLECTION_BYTES;
	g_pendingLuaHeapCollection = false;
}

void addTrackedLuaHeapBytes(ptrdiff_t delta) {
	if (delta < 0 && static_cast<size_t>(-delta) > g_trackedLuaHeapBytes) {
		throw std::runtime_error("[LuaHeapUsage] Tracked heap bytes underflow.");
	}
	g_trackedLuaHeapBytes = static_cast<size_t>(static_cast<ptrdiff_t>(g_trackedLuaHeapBytes) + delta);
	if (delta > 0) {
		requestLuaHeapCollectionIfNeeded();
	}
}

size_t trackedLuaHeapBytes() {
	return g_trackedLuaHeapBytes;
}

void enforceLuaHeapBudget() {
	const size_t totalRamUsedBytes = g_luaHeapUsageHooks.getBaseRamUsedBytes(g_luaHeapUsageHooks.context) + g_trackedLuaHeapBytes;
	if (!g_pendingLuaHeapCollection && totalRamUsedBytes < RAM_SIZE) {
		return;
	}
	collectTrackedLuaHeapBytesNow();
}

}
